### PR TITLE
Add FlyteExecConf and DockerExecConf to WorkflowConfiguration

### PR DESCRIPTION
**Pull request by narape**
# Hey, I just made a Pull Request!

## Description
This add some extra fields to `WorkflowConfiguration` to support Flyte execution configuration 
This PR docker and Flyte are not in the "same level" but the refactoring of Docker parameters will be dealt on a different PR.

```
Workflow: {
    componentId: String,
    workflowId: String,
    configuration: WorkflowConfiguration {
        workflowId: String,
        schedule: Schedule {
            expression: String
        }
        offset: Optional<String>,
        commitSha: Optional<String>,
        serviceAccount: Optional<String>,
        resources: List<String>,
        env: Map<String, String>,
        runningTimeout: Optional<Duration>,
        retryCondition: Optional<String>,
        dockerImage: Optional<String>,      <------- docker conf, we will move it later
        dockerArgs: Optional<List<String>>, <--+
        dockerTerminationLogging: boolean,  <--+
        flyteExecConf: Optional<FlyteExecConf> { <------ new flyte conf
            referenceId: FlyteIdentifier {
                resourceType: String,
                project: String,
                domain: String,
                name: String,    
                version: String,
            },
            inputFieldNames: Map<String, String>
        }
    }
}
```
## Motivation and Context
This is the start of the transition of making Styx able to execute launch plans on Flyte. 
To do so we need to modify our WorkflowConfiguration to specify this info.
Because this is so central to styx a lot of tests are needed to make sure that we are able to parse docker and flyte config style

## Have you tested this? If so, how?
 

I have included unit tests **and we will run a load test in staging**
## Checklist for PR author(s)

- [x] Changes are covered by unit test
- [x] All tests pass
- [x] Code coverage check passes
- [x] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [ ] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java [hunk 1]
```diff
@@ -267,7 +275,119 @@ public void shouldAcceptValidServiceAccount(String serviceAccount) {
...
+      identifierMutator.accept(referenceIdBuilder);
+      return confBuilder.flyteExecConf(
+          FlyteExecConfBuilder.from(flyteExecConf)
+              .referenceId(referenceIdBuilder.build())
+              .build())
+          .build();
+    }
+  }
+
+    private String limit(String msg, Object value, Object limit) {
```
Comment (sonjaer) [inline styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java:390]: I think this has wrong indentation

File: styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java [hunk 2]
```diff
@@ -261,17 +289,24 @@ public void tearDown() throws Exception {
...
+  private static WorkflowConfiguration[] configurations() {
+    return new WorkflowConfiguration[] {
+        FULL_WORKFLOW_CONFIGURATION, DOCKER_EXEC_WORKFLOW_CONFIGURATION, FLYTE_EXEC_WORKFLOW_CONFIGURATION
+    };
+  }
+
   @Test
-  public void shouldDeleteWorkflows() throws Exception {
+  @Parameters(method = "configurations")
+  public void shouldDeleteWorkflows(WorkflowConfiguration configuration) throws Exception {
```
Comment (sonjaer) [inline styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java:309]: Not exactly sure what the `@Parameters` does, but the parameter to this function doesn't seem to be used inside it

Comment (narape) [reply styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java:309]: Done

File: styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java [hunk 3]
```diff
@@ -222,10 +225,35 @@
...
   private static final WorkflowId WORKFLOW_ID = WorkflowId.create("dockerComp", "dockerEndpoint");
 
+  private static final String DOCKER_IMAGE = "gcr.io/foo/bar";
   private static final WorkflowConfiguration WORKFLOW_CONFIGURATION =
       WorkflowConfiguration.builder()
           .id(WORKFLOW_ID.id())
           .schedule(DAYS)
+          .dockerImage(DOCKER_IMAGE)
+          .build();
+  private static final WorkflowConfiguration DOCKER_EXEC_WORKFLOW_CONFIGURATION =
```
Comment (sonjaer) [inline styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java:235]: Is there a reason we don't use the TestData.java declaration of these variables?

**Comment by sonjaer** [comment]
Added changes to:
* Ignore flyte workflows from the TriggeringManager
* Disabling users posting backfills for flyte workflows
* Disabling users doing adhoc triggering of flyte workflows
and tested this. Not sure if it is possible to test more end to end that a flyte workflow can't currently be scheduled, as my knowledge on Styx isn't 100%.

File: styx-common/src/main/java/com/spotify/styx/model/WorkflowConfiguration.java [hunk 4]
```diff
@@ -42,10 +42,10 @@
 
   Optional<String> offset();
 
-  Optional<String> dockerImage();
-
   Optional<String> commitSha();
 
+  Optional<String> dockerImage();
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/model/WorkflowConfiguration.java:47]: What is the strategy for deprecating these?

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java [hunk 5]
```diff
@@ -66,15 +71,13 @@ public BasicWorkflowValidator(DockerImageValidator dockerImageValidator) {
     final List<String> e = new ArrayList<>();
 
     var componentId = workflowId.componentId();
-    if (componentId.isEmpty()) {
-      e.add("component id cannot be empty");
-    } else if (componentId.contains("#")) {
+
+    empty(e, "component id", workflowId.componentId());
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:75]: Should we rename this to something more descriptive? like checkIfEmpty or something

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java [hunk 6]
```diff
@@ -144,11 +137,97 @@ public BasicWorkflowValidator(DockerImageValidator dockerImageValidator) {
...
+    }
+  }
+
+  private void empty(List<String> errors, String fieldName, String value) {
+    if (value.isEmpty()) {
+      errors.add(fieldName + " cannot be empty");
+    }
+  }
+
+  private <T> void same(List<String> errors, String fieldName, T valueA, T valueB) {
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:228]: should we rename this one as well?

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java [hunk 7]
```diff
@@ -144,11 +137,97 @@ public BasicWorkflowValidator(DockerImageValidator dockerImageValidator) {
...
+  }
+
+  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+  private <T> void validateDockerParam(List<String> errors, String paramName, Optional<T> valueA, Optional<T> valueB,
+                                       BiConsumer<List<String>, T> validateFunction) {
+    valueA.ifPresent(a -> validateFunction.accept(errors, a));
+    valueB.ifPresent(b -> validateFunction.accept(errors, b));
+
+    valueA.ifPresent(a ->
+        valueB.ifPresent(b -> same(errors, paramName, a, b))
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:183]: Do we expect that both of these fields will be present and the same?

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java [hunk 8]
```diff
@@ -144,11 +137,97 @@ public BasicWorkflowValidator(DockerImageValidator dockerImageValidator) {
...
   }
 
   private static boolean validateServiceAccount(String serviceAccount) {
     var matcher = VALID_EMAIL_ADDRESS_REGEX.matcher(serviceAccount);
     return matcher.matches();
   }
+
+  private void validateDockerExecConf(List<String> errors, WorkflowConfiguration config) {
+    validateDockerParam(errors, "dockerImage", config.dockerImage(),
+        config.dockerExecConf().flatMap(DockerExecConf::dockerImage),
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:154]: Do we expect that both of these dockerImages should be present and the same?

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java [hunk 9]
```diff
@@ -144,11 +137,97 @@ public BasicWorkflowValidator(DockerImageValidator dockerImageValidator) {
...
+    //              .dockerImage("busybox")
+    //              .dockerArgs(List.of("x", "y"))
+    //              .dockerTerminationLogging(true)
+    //              .build())
+    //          .build();
+    // will fail as default at upper level is false but inside dockerExecConf is set to true
+  }
+
+  @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+  private <T> void validateDockerParam(List<String> errors, String paramName, Optional<T> valueA, Optional<T> valueB,
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:177]: I kind of feel that this abstraction creates unnecessary complication to the code

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java (hunk 5 above)
Comment (ckiosidis) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:75]: or shouldNotBeEmpty

File: styx-common/src/main/java/com/spotify/styx/testdata/TestData.java [hunk 10]
```diff
@@ -140,6 +143,37 @@
           .retryCondition("#exitCode == 1 && (#tries < 3 || #consecutiveFailures < 4) && #triggerType == \"natural\"")
           .build();
 
+  public static final WorkflowConfiguration DOCKER_EXEC_WORKFLOW_CONFIGURATION =
+      WorkflowConfiguration.builder()
+          .id("styx.TestEndpoint")
+          .commitSha(VALID_SHA)
+          .schedule(DAYS)
+          .serviceAccount("foo@bar.baz.quux")
+          .dockerExecConf(new DockerExecConfBuilder()
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/testdata/TestData.java:152]: Why do we have a 
WorkflowConfiguration.builder(), but then a `new DockerExecConfBuilder()` 

Should we be consistent with what we use?

File: styx-common/src/main/java/com/spotify/styx/testdata/TestData.java [hunk 11]
```diff
@@ -140,6 +143,37 @@
...
+              .build())
+          .build();
+
+  public static final WorkflowConfiguration FLYTE_WORKFLOW_CONFIGURATION =
+      WorkflowConfiguration.builder()
+          .id("styx.TestEndpoint")
+          .commitSha(VALID_SHA)
+          .schedule(DAYS)
+          .serviceAccount("foo@bar.baz.quux")
+          .flyteExecConf(new FlyteExecConfBuilder()
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/testdata/TestData.java:165]: Let's make the Builder usage consistent with the WorkflowConfiguration.builder()

File: styx-common/src/main/java/com/spotify/styx/testdata/TestData.java [hunk 12]
```diff
@@ -140,6 +143,37 @@
...
+              .referenceId(new FlyteIdentifierBuilder()
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/testdata/TestData.java:166]: Let's make the Builder usage consistent with the WorkflowConfiguration.builder()

File: styx-common/src/test/java/com/spotify/styx/model/WorkflowConfigurationTest.java [hunk 13]
```diff
@@ -76,4 +86,116 @@ public void shouldReturnDefaultOffset() {
...
+  }
+
+  public static class ParseJsonArgsProvider {
+
+    private ParseJsonArgsProvider(){
+      // Prevents instantiation
+    }
+
+    @SuppressWarnings("unused")
+    public static Object[] provideDockerCentredConf() {
```
Comment (ckiosidis) [inline styx-common/src/test/java/com/spotify/styx/model/WorkflowConfigurationTest.java:107]: do we mean dockerExecConf ?

File: styx-common/src/test/java/com/spotify/styx/model/WorkflowConfigurationTest.java [hunk 14]
```diff
@@ -76,4 +86,116 @@ public void shouldReturnDefaultOffset() {
...
+             + "\"service_account\":\"foo@project.com\","
+             + "\"resources\":[\"resource\"],"
+             + "\"env\":{\"STYX_FOO\":\"bar\"},"
+             + "\"running_timeout\":10800.000000000,"
+             + "\"retry_condition\":\"true\","
+             + execConf
+             + "}";
+    }
+  }
 }
```
Comment (ckiosidis) [inline styx-common/src/test/java/com/spotify/styx/model/WorkflowConfigurationTest.java:201]: missing new line

File: styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java [hunk 15]
```diff
@@ -267,6 +275,118 @@ public void shouldAcceptValidServiceAccount(String serviceAccount) {
...
+        sut.validateWorkflow(Workflow.create("test", DOCKER_EXEC_WORKFLOW_CONFIGURATION)),
+        empty()
+    );
+  }
+
+  @Test
+  public void shouldRejectInvalidImageInDockerExecConf() {
+    when(dockerImageValidator.validateImageReference(anyString())).thenReturn(Set.of("foo", "bar"));
+
+    var errors = sut.validateWorkflow(Workflow.create("test", DOCKER_EXEC_WORKFLOW_CONFIGURATION));
```
Comment (ckiosidis) [inline styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java:290]: Should we instead create an INVALID_DOCKER_EXEC_WORKFLOW_CONFIGURATION, and make this test a bit more explicit?

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java [hunk 16]
```diff
@@ -152,7 +152,8 @@ private void tryTriggering(Workflow workflow,
   private void tryTriggering0(Workflow workflow,
                                  TriggerInstantSpec instantSpec,
                                  Set<WorkflowId> enabledWorkflows) {
-    if (enabledWorkflows.contains(workflow.id())) {
+    // Ignore flyte workflows
+    if (enabledWorkflows.contains(workflow.id()) && workflow.configuration().flyteExecConf().isEmpty()) {
```
Comment (ckiosidis) [inline styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:156]: I think it makes sense to remove the flyteExecConf from this if check, and instead do an early return with it's own debug log

File: styx-test/src/test/java/com/spotify/styx/storage/DatastoreEmulatorTest.java [hunk 17]
```diff
@@ -66,11 +64,12 @@ public void testEmulator(Datastore client) {
...
     var emulator = new DatastoreEmulator();
-    exception.expect(AssertionError.class);
-    exception.expectMessage(Matchers.startsWith("Not using gcloud sdk datastore emulator"));
-    emulator.before();
+
+    var exception = Assert.assertThrows(AssertionError.class, emulator::before);
+
+    assertThat(exception.getMessage(), Matchers.startsWith("Not using gcloud sdk datastore emulator"));
   }
 }
```
Comment (ckiosidis) [inline styx-test/src/test/java/com/spotify/styx/storage/DatastoreEmulatorTest.java:75]: new line

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java [hunk 18]
```diff
@@ -65,7 +66,8 @@
  */
 class TriggerManager implements Closeable {
 
-  private static final Logger LOG = LoggerFactory.getLogger(TriggerManager.class);
+  private static final Logger DEFAULT_LOG = LoggerFactory.getLogger(TriggerManager.class);
+  private final Logger log;
```
Comment (ckiosidis) [inline styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:70]: we don't need a different logger here. Let's reuse the first one.

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java [hunk 19]
```diff
@@ -152,19 +163,23 @@ private void tryTriggering(Workflow workflow,
...
                                  TriggerInstantSpec instantSpec,
                                  Set<WorkflowId> enabledWorkflows) {
-    // Ignore flyte workflows
-    if (enabledWorkflows.contains(workflow.id()) && workflow.configuration().flyteExecConf().isEmpty()) {
+
+    boolean isFlyteWorkflow = workflow.configuration().flyteExecConf().isPresent();
+
+    if (isFlyteWorkflow) {
+      log.info("Skip triggering flyte workflow");
+    } else if (enabledWorkflows.contains(workflow.id())) {
```
Comment (ckiosidis) [inline styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:171]: It's better here to do
```
    if (isFlyteWorkflow) {
      log.info("Skip triggering flyte workflow");
      return;
    }

    if (enabledWorkflows.contains(workflow.id())) {...
```

this way we separate from the other (unrelated logic), and we return early and safe

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java (hunk 18 above)
Comment (sonjaer) [reply styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:70]: I added this to be able to pass a mocked logger from the test class to verify that the logging happened

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java (hunk 19 above)
Comment (sonjaer) [reply styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:171]: I did it this way since I want this code to execute: https://github.com/spotify/styx/pull/830/commits/5fbba776fb6f5edc968d52fecc7d8001ec8edf44#diff-8681f2f92e0c50ca9a99b3fc6ee3e4a2R196 So that the next trigger date updates

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java (hunk 5 above)
Comment (narape) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:75]: I agreed that those are better names but I was being consistent with `limit`. We can make this change and limit in a following PR

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java (hunk 8 above)
Comment (narape) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:154]: Yes it is confusing. I will change it to be a clear cut old style or clear cut new style in flyte or docker

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java (hunk 18 above)
Comment (narape) [reply styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:70]: Usually verifying logging is not needed. An exception case would be that logging is the interface between systems, like our audit logging

File: styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java (hunk 3 above)
Comment (narape) [reply styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java:235]: No really, I will move code overthere

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java (hunk 18 above)
Comment (sonjaer) [reply styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:70]: Makes sense, I can revert that last commit

File: styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java (hunk 19 above)
Comment (ckiosidis) [reply styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:171]: Do we want this to happen? and run the `updateNextNaturalTrigger` regardless if it will fail next time?

Comment (sonjaer) [reply styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:171]: I think so. I assume we want to handle it in the same way as disabled workflows. Otherwise, I think that once we enable the ability to run flyte workflows that each workflow would backfill from the date it was created until we enable the functionality. Not that this would maybe be a huge problem, but seems better to not cause them to backfill.

Comment (sonjaer) [reply styx-scheduler-service/src/main/java/com/spotify/styx/TriggerManager.java:171]: This is my understanding of what happens. But would be great with a second pair of eyes.

File: styx-common/src/main/java/com/spotify/styx/model/WorkflowConfiguration.java (hunk 4 above)
Comment (narape) [reply styx-common/src/main/java/com/spotify/styx/model/WorkflowConfiguration.java:47]: Step 1: Change model to allow both. This PR
Step 2: Remove old methods, keep tests that added in this PR to make sure that we still can read old json and read old entities in Datastore but converting old docker params to new DockerExecConf.

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java [hunk 20]
```diff
@@ -149,45 +147,33 @@ private static boolean validateServiceAccount(String serviceAccount) {
...
-    //              .dockerTerminationLogging(true)
-    //              .build())
-    //          .build();
-    // will fail as default at upper level is false but inside dockerExecConf is set to true
+  private void validateConflictingExecConf(List<String> e, WorkflowConfiguration cfg) {
+    var hasOldStyleDockerConf = cfg.dockerImage().isPresent() || cfg.dockerArgs().isPresent();
+    var hasNewStyleDockerConf = cfg.dockerExecConf().isPresent();
+
+    if (hasNewStyleDockerConf && hasOldStyleDockerConf) {
+      e.add("configuration cannot docker parameters in new and old style");
```
Comment (sonjaer) [inline styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:155]: Nothing big, but missing `specify both`

File: styx-scheduler-service/src/test/java/com/spotify/styx/TriggerManagerTest.java [hunk 21]
```diff
@@ -54,23 +54,21 @@
...
 public class TriggerManagerTest {
 
   private static final Trigger NATURAL_TRIGGER = Trigger.natural();
 
-  private static Workflow WORKFLOW_DAILY =
+  private static final Workflow WORKFLOW_DAILY =
       Workflow.create("comp", FULL_WORKFLOW_CONFIGURATION);
 
-  private static Workflow FLYTE_WORKFLOW =
+  private static final Workflow FLYTE_WORKFLOW =
```
Comment (sonjaer) [inline styx-scheduler-service/src/test/java/com/spotify/styx/TriggerManagerTest.java:66]: This could be removed

File: styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java [hunk 22]
```diff
@@ -88,6 +103,8 @@
   public void setUp() throws Exception {
     MockitoAnnotations.initMocks(this);
     when(dockerImageValidator.validateImageReference(anyString())).thenReturn(Set.of());
+    when(dockerImageValidator.validateImageReference(eq(NOT_VALID_IMAGE)))
```
Comment (sonjaer) [inline styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java:106]: Nothing important, but do we need to mock this? Is that good practice? Or is there another way to do it?

File: styx-common/src/main/java/com/spotify/styx/model/WorkflowConfiguration.java (hunk 4 above)
Comment (ckiosidis) [reply styx-common/src/main/java/com/spotify/styx/model/WorkflowConfiguration.java:47]: 👍

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java (hunk 5 above)
Comment (ckiosidis) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:72]: cool 👍

File: styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java (hunk 22 above)
Comment (narape) [reply styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java:106]: Well I guess it depends to whom you ask. 
* Some people are against mocking because they say you should create interface for everything and then implement simple test classes implementation of those interfaces. 
* Other people are against creating "unnecessary" interfaces because that is what mocking libraries are for.  

I see that both are possible approaches but I find Mocking libraries are quite handy and I don't find the downsides of mocking that convincing.

To the question of do we need to mock this? I say that we want verify that we keep calling the validator and react to it as we should, if I comment by mistake the if block that calls the validator then I want a test to fail. So we need a test for that and we want to control when to fail and when not to fail. We could achieve this by mocking (as done in the PR) or we could have introduced a new interface and make the validator implement that interface and create a test class that implements that interface but it is certainly more work that just adding two lines of mocking.

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java [hunk 23]
```diff
@@ -149,45 +147,33 @@ private static boolean validateServiceAccount(String serviceAccount) {
...
+    }
+    if (cfg.flyteExecConf().isPresent() && (hasNewStyleDockerConf || hasOldStyleDockerConf)) {
+      e.add("configuration cannot specify both docker and flyte parameters");
```
Comment (ckiosidis) [inline styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:158]: shouldn't it be allowed to have flyteExecConf and dockerExecConf ?

Comment (narape) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:158]: I guess you are asking because for migrating workflows we said that we will switch workflows from Docker to Flyte. But this PR is not that. My intention with this PR is to have either docker or flyte but not both. Future PRs we will look on how to do the switch and I didn't want to mix concerns

File: styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java [hunk 24]
```diff
@@ -1131,8 +1105,13 @@ public void runInTransactionWithRetriesShouldPropagateDatastoreIOException() thr
...
     doThrow(cause).when(datastore).newTransaction();
-    exception.expect(is(cause));
-    storage.runInTransactionWithRetries(transactionFunction);
+
+    var exception = assertThrows(
+        DatastoreIOException.class,
+        () -> storage.runInTransactionWithRetries(transactionFunction)
+    );
+
+    assertThat(exception, is(cause));
```
Comment (ckiosidis) [inline styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java:1114]: these tests look a bit too rigid, but good refactoring

File: styx-scheduler-service/src/main/java/com/spotify/styx/StateInitializingTrigger.java [hunk 25]
```diff
@@ -50,13 +51,18 @@
   @Override
   public void event(Workflow workflow, Trigger trigger, Instant instant,
                     TriggerParameters parameters) {
-    if (workflow.configuration().dockerImage().isEmpty()) {
+    final var configuration = workflow.configuration();
+    if (configuration.flyteExecConf().isPresent()) {
+      LOG.info("{} is a Flyte Workflow, skipping as not supported at the moment", workflow.id());
+      return;
+    } else if (configuration.dockerImage().isEmpty()
```
Comment (ckiosidis) [inline styx-scheduler-service/src/main/java/com/spotify/styx/StateInitializingTrigger.java:58]: The flyteExecConf and dockerImage look unrelated at first.
Can we add a comment here describing the relation and why we need them in a else if block, instead of two separate early returns?
It should help us moving on when working with the code.

File: styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java (hunk 24 above)
Comment (narape) [reply styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java:1114]: This was more a "Boy Scout rule" change as the way for asserting for exception using `@Rule` is now deprecated in favor of `assertThrows`. Functionally this test is as strict as the old version, just using a not deprecated API

File: styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java [hunk 26]
```diff
@@ -123,19 +140,37 @@ public void shouldInjectTriggerExecutionEventWithBackfillTrigger() throws Except
   }
 
   @Test
-  public void shouldDoNothingIfDockerImageMissing() {
-    final WorkflowConfiguration configuration =
+  @Parameters(method = "noDockerImageConfigurations")
+  public void shouldDoNothingIfDockerImageMissing(WorkflowConfiguration configuration) {
```
Comment (ckiosidis) [inline styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java:144]: What is the benefit of having this method Parameterized instead of having a static generation of the configuration, or defining it in the beginning of the method?

File: styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java [hunk 27]
```diff
@@ -150,12 +185,26 @@ public void shouldCreateWorkflowInstanceParameter() throws Exception {
...
     return WorkflowConfiguration.builder()
         .id("styx.TestEndpoint")
         .schedule(schedule)
         .dockerImage("busybox")
         .dockerArgs(ImmutableList.copyOf(args))
         .build();
   }
+
+  @Test
+  public void shouldWrapIsClosedExceptionIntoRuntime() throws Exception {
```
Comment (ckiosidis) [inline styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java:198]: what are we testing in this test? is it something we added in this commit?

File: styx-scheduler-service/src/main/java/com/spotify/styx/StateInitializingTrigger.java (hunk 25 above)
Comment (narape) [reply styx-scheduler-service/src/main/java/com/spotify/styx/StateInitializingTrigger.java:58]: Aha, yes good point. My tests didn't pass until I did the else if but I didn't document my finding

File: styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java (hunk 22 above)
Comment (ckiosidis) [reply styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java:106]: Mocking here is ok if we have others tests where we test the behaviour of the `dockerImageValidator.validateImageReference`.
Since this test only makes sure that the method is called, and not that the validation logic behaves properly.
Here are the other tests: https://github.com/spotify/styx/blob/master/styx-common/src/test/java/com/spotify/styx/util/DockerImageValidatorTest.java

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java (hunk 23 above)
Comment (ckiosidis) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:158]: right i hear you.
So what am I missing here?
We don't care about the migration/switch here.
So we should have 2 valid cases:
- either it's the old workflow configuration.
- or it's Flyte
It seems to me that the if-statement will trigger whenever Flyte is available.

File: styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java (hunk 24 above)
Comment (ckiosidis) [reply styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java:1114]: yes I agree. I mean that the test itself, not the refactoring

File: styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java (hunk 26 above)
Comment (sonjaer) [reply styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java:144]: The test is running for multiple configurations

File: styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java (hunk 22 above)
Comment (sonjaer) [reply styx-common/src/test/java/com/spotify/styx/util/BasicWorkflowValidatorTest.java:106]: Ok thank you 👍  Was mostly asking out of curiosity.

File: styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java (hunk 27 above)
Comment (narape) [reply styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java:198]: No it was a hole is our test coverage

File: styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java (hunk 26 above)
Comment (ckiosidis) [reply styx-scheduler-service/src/test/java/com/spotify/styx/StateInitializingTriggerTest.java:144]: Ah right. Didn't notice it was 2 configurations. cool!

File: styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java (hunk 23 above)
Comment (ckiosidis) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:152]: Hm right. So it should be either the FlyteExecConf or the docker configuration. 👍

Comment (ckiosidis) [reply styx-common/src/main/java/com/spotify/styx/util/BasicWorkflowValidator.java:152]: Can we add a comment explaining that we dont need to store docker configuration with flyte, because it's stored on flyte itself?
